rust/ql/lib/codeql/rust/internal/PathResolution.qll

```diff
@@ -1563,6 +1667,41 @@ module TraitIsVisible<relevantTraitVisibleSig/2 relevantTraitVisible> {
   }
 }
 
+pragma[nomagic]
+private predicate isMacroExpansion(AstNode expansion, Path macroDefPath) {
+  exists(MacroCall mc |
+    expansion = mc.getMacroCallExpansion() and
+    macroDefPath = mc.getPath()
+  )
+  or
+  exists(ItemNode adt |
+    expansion = adt.(Adt).getDeriveMacroExpansion(_) and
+    macroDefPath = adt.getAttr("derive").getMeta().getPath()
+  )
+}
+
+pragma[nomagic]
+predicate isInMacroExpansion(Path macroDefPath, AstNode n) {
+  isMacroExpansion(n, macroDefPath)
+  or
+  isInMacroExpansion(macroDefPath, n.getParentNode())
```

Comment:
I don't understand why `n.getParentNode()` is sufficient? I would've expected something like `n.getParentNode*()` as the `$crate` path can appear anywhere in the expansion?

Reply:
The predicate is already recursive.